### PR TITLE
Add support for SAMD20 Xplained Pro Board

### PR DESCRIPTION
**Pull request by sknsean**
I'm in the process of porting Zephyr to a custom board, MCU ATSAMD20E18.
First step is to get samd20 support included :-)

### CODE REVIEW

**Comment by pfalcon** [comment]
@nzmichaelh : FYI

**Comment by pfalcon** [comment]
@jpfaff : FYI

**Comment by nzmichaelh** [comment]
Thanks @pfalcon, LGTM.

**Comment by mnkp** [comment]
As far as I understand Zephyr development process no single commit should break the build. This means that commits:
- ext: atmel: rename samd folder to samd21
- boards: arduino_zero: use samd21 soc series
- arch: arm: soc: atmel_sam0: rename samd folder to samd21

should be combined into one: "arch: atmel_sam0: rename samd SoC series to samd21. While files in different folders need to be touched this is a one atomic change that we are doing.

File: ext/hal/atmel/asf/sam0/include/samd20/sam.h
Comment (mnkp) [inline ext/hal/atmel/asf/sam0/include/samd20/sam.h]: This file is not required, we could remove it from commit -> less clutter. Other existing imports from Atmel ASF do not include this file.

File: ext/hal/atmel/asf/sam0/include/samd20/samd20.h
Comment (mnkp) [inline ext/hal/atmel/asf/sam0/include/samd20/samd20.h]: This file is not required.

File: ext/hal/atmel/asf/sam0/include/samd20/system_samd20.h
Comment (mnkp) [inline ext/hal/atmel/asf/sam0/include/samd20/system_samd20.h]: This file is not required.

**Review by mnkp — CHANGES_REQUESTED** [review]
ASF library files have windows style endings "\r\n", we should convert it to Unix "\n". As far as I know checkpatch does not run on ext/ folder so such errors are not caught automatically. We probably should fix checkpatch.

**Comment by sknsean** [comment]
Squashed the 3 move commits
Fixed line endings :-)

**Comment by mnkp** [comment]
Thanks! LGTM, with a minor remark. The comment to the first commit "Make for room for other samd series" could be improved. Maybe something like "Atmel SAMD21 series was classified too broadly as SAMD. This patch names it correctly to make room for other members of SAMD series".

**Comment by sknsean** [comment]
@mnkp agreed fixed commit msg :-)

**Comment by mnkp** [comment]
Sorry, I was not clear enough. The commit's first line ""arch: atmel_sam0: rename samd SoC series to samd21" was OK. It's short, to the point and said what the commit did. Only the body of the commit message was not very descriptive. I guess you know already that the first comment line in git is special. To sum it up I would write a commit message like this

arch: atmel_sam0: rename samd SoC series to samd21

Atmel SAMD21 series was classified too broadly as SAMD. This patch names it correctly to make room for other SAMD series.

**Comment by sknsean** [comment]
Okay :-)
I removed the first line as it changed files in a lot of other directories than arch/atmel_sam0..
I'll re-add that line and fix some issues after the spi driver was merged.

**Comment by sknsean** [comment]
@mnkp fixed and green again :-)

**Review by mnkp — APPROVED** [review]

**Comment by sknsean** [comment]
@nashif @galak What do i need to do to get this merged?
I'm using this for my own custom board, do I need to buy the evalkit with the CPU and make a complete PR for support for that?

**Comment by galak** [comment]
> @nashif @galak What do i need to do to get this merged?
> I'm using this for my own custom board, do I need to buy the evalkit with the CPU and make a complete PR for support for that?

We need the code to be testable if we accept it upstream, so having a CPU/SoC & board port that uses the code is something of a prereq for merging.  I'm find with merging the rename patch right now if you want to pull that out into its own PR.

**Comment by sknsean** [comment]
Great will buy the evalkit :-)
And move the rename patch to it's own PR

**Comment by galak** [comment]
I'm good with this, but will wait on SoC & board port before accepting.

**Comment by sknsean** [comment]
@galak and @nzmichaelh
Please take a look :-)

File: dts/arm/atmel/samd.dtsi
Comment (nzmichaelh) [inline dts/arm/atmel/samd.dtsi]: nit: Atmel now have a SAMD51 series which seems to be a SAM0 with a Cortex M4.  I wouldn't worry about it now, but we may need to rename this in the future.

File: arch/arm/soc/atmel_sam0/samd20/Kconfig.defconfig.series
Comment (nzmichaelh) [inline arch/arm/soc/atmel_sam0/samd20/Kconfig.defconfig.series]: optional: move this to the generic SAM0 configuration

File: arch/arm/soc/atmel_sam0/samd20/Kconfig.series
Comment (nzmichaelh) [inline arch/arm/soc/atmel_sam0/samd20/Kconfig.series]: "Atmel SAMD20 MCU series"

File: arch/arm/soc/atmel_sam0/samd20/soc.c
Comment (nzmichaelh) [inline arch/arm/soc/atmel_sam0/samd20/soc.c]: WDYT about making a atmel_sam0/common directory and putting this in there instead?  This is how the atmel_sam/ common code is handled.

File: arch/arm/soc/atmel_sam0/samd20/soc_pinmap.h
Comment (nzmichaelh) [inline arch/arm/soc/atmel_sam0/samd20/soc_pinmap.h]: Note that this has moved to pinmux.c.  Consider doing the same in a follow up CL.

File: boards/arm/atsamd20_xpro/atsamd20_xpro.dts
Comment (nzmichaelh) [inline boards/arm/atsamd20_xpro/atsamd20_xpro.dts]: Huh, is this correct?  I thought the I2C port was also a SERCOM mode...

File: boards/arm/atsamd20_xpro/support/openocd.cfg
Comment (nzmichaelh) [inline boards/arm/atsamd20_xpro/support/openocd.cfg]: Do these need updating?

File: boards/arm/atsamd20_xpro/doc/atsamd20_xpro.rst
Comment (dbkinder) [inline boards/arm/atsamd20_xpro/doc/atsamd20_xpro.rst]: Change ``Including`` to ``The kit includes``

**Review by dbkinder — CHANGES_REQUESTED** [review]
One tweak, otherwise looks fine.  Thanks.

File: arch/arm/soc/atmel_sam0/samd20/soc.c
Comment (sknsean) [reply arch/arm/soc/atmel_sam0/samd20/soc.c]: They are not 100% identical, if we choose to add a common directory, then we would have to deal with some if defs

File: boards/arm/atsamd20_xpro/atsamd20_xpro.dts
Comment (sknsean) [reply boards/arm/atsamd20_xpro/atsamd20_xpro.dts]: are we still missing an i2c driver for the sam0?

Comment (henrikbrixandersen) [reply boards/arm/atsamd20_xpro/atsamd20_xpro.dts]: @sknsean Corrrect. There is no in-tree TWI driver for the sam0 series yet.

Comment (sknsean) [reply boards/arm/atsamd20_xpro/atsamd20_xpro.dts]: Great one more thing on the TODO list :-)

File: boards/arm/atsamd20_xpro/atsamd20_xpro.dts
Comment (henrikbrixandersen) [inline boards/arm/atsamd20_xpro/atsamd20_xpro.dts]: Please sync up with the changes in 9968cda453ac7a91d513b6a50817c926c3fe5cc6.

File: boards/arm/atsamd20_xpro/doc/atsamd20_xpro.rst
Comment (henrikbrixandersen) [inline boards/arm/atsamd20_xpro/doc/atsamd20_xpro.rst]: Please sync up with be52e3c456f0694fb98a4704860102d61048fd05.

**Review by henrikbrixandersen — DISMISSED** [review]

File: boards/arm/atsamd20_xpro/atsamd20_xpro.dts
Comment (sknsean) [reply boards/arm/atsamd20_xpro/atsamd20_xpro.dts]: :+1:

File: boards/arm/atsamd20_xpro/doc/atsamd20_xpro.rst
Comment (sknsean) [reply boards/arm/atsamd20_xpro/doc/atsamd20_xpro.rst]: :+1:

**Review by dbkinder — APPROVED** [review]
+1 for doc

File: dts/arm/atmel/samd.dtsi
Comment (mnkp) [reply dts/arm/atmel/samd.dtsi]: We probably shouldn't be doing it. SAMD family has the following series: D09, D10, D11, D20, D21, DA1, HA1 and they are all quite different. E.g several D10 peripherals are located at different addresses than D21/D20. While it's true that D20 and D21 look fairly similar if we try to maintain one common file it will very quickly get messy. Even at present this file contains an entry for USB0 peripheral which is missing on D20 chip.

Let's maybe keep one dtsi file per series and only ensure that they are easy to compare with diff.

**Review by mnkp — DISMISSED** [review]

File: arch/arm/soc/atmel_sam0/Kconfig
Comment (mnkp) [inline arch/arm/soc/atmel_sam0/Kconfig]: Settings in this file should be applicable to the whole SAM0 family, not only SAM Dxx series. SAM0 family has following members: B11, C20, C21, D09, D10, D11, D20, D21, DA1, HA1, L21, L22, R21, R30. Some of them don't have 8 MHz RC oscillator.

Comment (sknsean) [reply arch/arm/soc/atmel_sam0/Kconfig]: Could we call it SAMD2X then?

Comment (mnkp) [reply arch/arm/soc/atmel_sam0/Kconfig]: If we are able to reduce the number of configuration files by recognizing similarities between different chips that's probably good. We would however need to ensure that our implementation is correct and there are no (not yet implemented) series that will break it. Looking at the datasheets of D09, D10, D11, D20, D21, DA1 series they all seem to have identical clock subsystem. So using SAMD common clock options may be an option. But we would need to move it to some other Kconfig file as this one is meant for the whole SAM0 family. Also we would need to ensure that if we ever add e.g. B11 series the SAMD options will not be sourced.

That said this is not how other arch/arm/soc/* families/series are configured. So far every series has its own independent configuration files. It would be best to ask @galak or @MaureenHelm what's the best way to proceed.

**Comment by galak** [comment]
Still want to try and get this merged?

**Comment by sknsean** [comment]
Yes that’s why the PR is still open...
It seems like a lot of people had their say, but I still don’t know who the maintainer is and thereby I don’t know what needs to be fixed before it can be merged. 
The purpose of this was not to fix the bad handling/the whole world of multiple sam0 SoCs.

So please provide a list of what needs to be done

File: arch/arm/soc/atmel_sam0/Kconfig
Comment (galak) [reply arch/arm/soc/atmel_sam0/Kconfig]: I've cleaned this up, as we see new SoCs added and new patterns we can tweak things.  for now I added a Kconfig.samd2x for that stuff.

**Comment by galak** [comment]
> Yes that’s why the PR is still open...
> It seems like a lot of people had their say, but I still don’t know who the maintainer is and thereby I don’t know what needs to be fixed before it can be merged.
> The purpose of this was not to fix the bad handling/the whole world of multiple sam0 SoCs.
> 
> So please provide a list of what needs to be done

I merged down some of the commits, and updating a few things to sync with the latest upstream changes.  Will review to see if I have any other comments, while it runs CI checks.

**Comment by skny-prevas** [comment]
Great, thanks :+1:

**Comment by sknsean** [comment]
Great thanks :+1:

**Review by galak — APPROVED** [review]

**Pull request merged**
